divi/src/activemasternode.cpp

```diff
@@ -171,27 +171,9 @@ bool CActiveMasternode::SendMasternodePing(std::string& errorMessage)
 
 bool CActiveMasternode::Register(CMasternodeBroadcast &mnb, bool deferRelay)
 {
-    int nDoS = 0;
-    if(!mnb.CheckAndUpdate(nDoS) || !mnb.CheckInputsAndAdd(nDoS))
+    if (!mnodeman.ProcessBroadcast(nullptr, mnb))
         return false;
 
-    addrman.Add(CAddress(mnb.addr), CNetAddr("127.0.0.1"), 2 * 60 * 60);
-
-    auto mnp = mnb.lastPing;
-    mnodeman.mapSeenMasternodePing.insert(std::make_pair(mnp.GetHash(), mnp));
```

Comment:
masternodeman's ProcessBroadcast does not in fact record the ping for the broadcast inside the mapSeenMasternodePing. Shouldn't it happen then after the function call (on success, of course)?

Reply:
Yes, this bit was one of the differences I noticed between the three places unified in this PR.  I thought about adding it to the new shared `ProcessBroadcast` function, but then went for not doing it (but with no particularly strong reason).

As far as my current understanding goes, adding the ping is not necessary, since we never actually relay its `inv` and thus also won't need to answer any queries for it from peers (instead it gets relayed and processed directly with the `mnb` message).  Furthermore, if it were necessary, then the old code would be broken as well, since only the node actually performing the start used to insert it into `mapSeenMasternodePing` (and not every node processing the broadcast message).

But I can also see the argument that it seems cleaner / safer / more straight-forward to add the ping in, since it is a ping we've seen after all.  In that case, we should add it to `ProcessBroadcast`.  Let me know if that's the option you prefer and I'm happy to do that.

Reply:
I'm not advocating for moving the ping into the ProcessBroadcast function. More precisely, I'm merely indicating that during registration of a broadcast, the ping is recorded in order to avoid any confusion about the ping having been seen previously. Seems like it would lead to some communication confusion if I receive an unknown ping (that belongs to my masternode) but might somehow query my peers for it? Sounds like it would sort itself out but I don't like to make wagers like that haha

Reply:
I think with that change this is ready to be merged

Reply:
I've done an update now that adds the ping to `mapSeenMasternodePing` in `ProcessBroadcast`.  This is still slightly different from the old behaviour, but I don't see why processing should differ for this detail between starting one's own masternode and processing a message from the network.  (But if you have a good reason, please let me know!)